### ISSUE TITLE
distro-sync as default behavior?

### ISSUE THREAD

**Issue by kparal** [rpm-software-management/dnf-plugin-system-upgrade#10]
Please enable distro-sync mode by default. One of the major hopes I had for dnf-plugin-system-upgrade was to have finally functional distro-sync mode. The lack of distro-sync caused us endless issues with fedup ugprades. Maintainers can't be expected to keep upgrade path when Branched is frozen towards the release. We need to sync to Branched versions.

This has been discussed here [1] and people seemed to agree that distro-sync is the better default.

Also, dnf now resolves broken deps differently then fedup. Fedup used to download the target packages and force-install them, even if deps were broken. DNF now skips packages recursively, if they are broken. The end result is that if there's an upgrade path inconsistency in one of major system packages (like systemd, gtk, etc), people might end up with half system upgraded and half system old, because the packages were skipped due to broken deps. I don't see any other way to prevent this except for distro-sync mode. We need to make upgrades more reliable.

Please change `--distro-sync` option to something like `--no-distro-sync` option and set distro-sync as default. Thanks.

[1] http://lists.rpm.org/pipermail/rpm-ecosystem/Week-of-Mon-20150727/000180.html

**Comment by jsilhan** [comment]
please don't do this feature until we make `dnf distro-sync` without removal of packages
https://bugzilla.redhat.com/show_bug.cgi?id=1244175
https://bugzilla.redhat.com/show_bug.cgi?id=1259175

**Comment by kparal** [comment]
Let's add this one to @jsilhan's list as well:
https://bugzilla.redhat.com/show_bug.cgi?id=1260989

Now let's see how real system works with just "update" (no distro sync):
https://gist.github.com/kparal/c23e05cef084ce87c3ea
There are many many packages in "skipped due to broken dependencies" section. A large part of my system would go unupgraded, I don't suppose it would work properly after that. Unfortunately, many of our users will not study this output (or won't understand it), they'll just hit yes and end up with a broken system.

I hope this would get much better with distro-sync, because a newer package on the current system would get downgraded, instead of holding many other packages back. But it seems I can't provide a matching distro-sync output due to some of the dnf issues mentioned (and maybe there are other issues as well):

```
$ sudo dnf system-upgrade download --releasever=23 --distro-sync
Failed to synchronize cache for repo 'rpmfusion-nonfree-updates-testing' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=nonfree-fedora-updates-testing-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-free-updates-testing' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=free-fedora-updates-testing-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-free-updates' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=free-fedora-updates-released-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-nonfree-updates' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=nonfree-fedora-updates-released-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-free' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=free-fedora-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-nonfree' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=nonfree-fedora-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Last metadata expiration check performed 0:15:11 ago on Tue Sep  8 14:41:52 2015.
Error: package ffmpeg-libs-2.6.4-1.fc22.x86_64 requires libhogweed.so.2()(64bit), but none of the providers can be installed.
package python-solv-0.6.11-2.fc22.x86_64 requires libsolv(x86-64) = 0.6.11-2.fc22, but none of the providers can be installed.
package ffmpeg-2.6.4-1.fc22.x86_64 requires libavcodec.so.56()(64bit), but none of the providers can be installed.
package libavdevice-2.6.4-1.fc22.x86_64 requires libavcodec.so.56()(64bit), but none of the providers can be installed.
package vlc-core-2.2.1-6.fc22.x86_64 requires libavcodec.so.56()(64bit), but none of the providers can be installed.
package vlc-2.2.1-6.fc22.x86_64 requires libvlccore.so.8()(64bit), but none of the providers can be installed
(try to add '--allowerasing' to command line to replace conflicting packages)
```

```
$ sudo dnf system-upgrade download --releasever=23 --distro-sync --allowerasing
Failed to synchronize cache for repo 'rpmfusion-nonfree-updates-testing' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=nonfree-fedora-updates-testing-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-free-updates-testing' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=free-fedora-updates-testing-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-free-updates' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=free-fedora-updates-released-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-nonfree-updates' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=nonfree-fedora-updates-released-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-free' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=free-fedora-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Failed to synchronize cache for repo 'rpmfusion-nonfree' from 'http://mirrors.rpmfusion.org/mirrorlist?repo=nonfree-fedora-23&arch=x86_64': Cannot prepare internal mirrorlist: No URLs in mirrorlist, disabling.
Last metadata expiration check performed 0:15:33 ago on Tue Sep  8 14:41:52 2015.
Dependencies resolved.
Error: The operation would result in removing the booted kernel: kernel-core-4.1.6-200.fc22.x86_64.
```

**Comment by wgwoods** [comment]
> This has been discussed here [1] and people seemed to agree that distro-sync is the better default.

"people seemed to agree" is not sufficient for me to change the default behavior of the upgrade tool.

This is a policy decision more than a technical issue; I'm going to need this to go through FESCo first.

**Comment by johnp** [comment]
@kparal : The ffmpeg/vlc issue stems from rpmfusion not having rebuild their packages for f23:
https://bugzilla.rpmfusion.org/show_bug.cgi?id=3733

My testing also showed a subsequent NameError in dnf:

```
$ dnf system-upgrade download --releasever=23 --enablerepo=updates-testing --distro-sync
Last metadata expiration check performed 0:00:00 ago on Tue Sep  8 20:22:31 2015.
Error: package ffmpeg-libs-2.6.4-1.fc22.x86_64 requires libhogweed.so.2()(64bit), but none of the providers can be installed.
package ffmpeg-2.6.4-1.fc22.x86_64 requires libavcodec.so.56()(64bit), but none of the providers can be installed.
package libavdevice-2.6.4-1.fc22.x86_64 requires libavcodec.so.56()(64bit), but none of the providers can be installed.
package libvdpau-va-gl-0.3.4-6.fc22.x86_64 requires libswscale.so.3()(64bit), but none of the providers can be installed.
package vlc-core-2.2.1-6.fc22.x86_64 requires libavcodec.so.56()(64bit), but none of the providers can be installed.
package vlc-2.2.1-6.fc22.x86_64 requires libvlccore.so.8()(64bit), but none of the providers can be installed.
package vlc-extras-2.2.1-6.fc22.x86_64 requires libvlccore.so.8()(64bit), but none of the providers can be installed
Traceback (most recent call last):
  File "/bin/dnf", line 36, in <module>
    main.user_main(sys.argv[1:], exit_code=True)
  File "/usr/lib/python2.7/site-packages/dnf/cli/main.py", line 195, in user_main
    errcode = main(args)
  File "/usr/lib/python2.7/site-packages/dnf/cli/main.py", line 94, in main
    demands = self.cli.demands
NameError: global name 'self' is not defined
```

**Comment by kparal** [comment]
> "people seemed to agree" is not sufficient for me to change the default behavior of the upgrade tool.
> This is a policy decision more than a technical issue; I'm going to need this to go through FESCo first.

OK, let's get technical then. Currently, in my experience, system-upgrade fares much worse than fedup does. system-upgrade wants to install/update 1763 packages and leave 263 packages not updated (including critical ones like `cups`, `perl`,`boost-*`, `gnutls` and others). I'm afraid the upgraded system would be very broken, and it certainly wouldn't be considered fully upgraded.

Let's see fedup:

```
fedup.yum INFO: buildTransaction returned 1
fedup.yum INFO:     python-solv-0.6.11-2.fc22.x86_64 requires libsolv(x86-64) = 0.6.11-2.fc22
fedup.yum INFO:     audit-libs-python3-2.4.3-1.fc23.x86_64 requires audit = 2.4.3-1.fc23
fedup.yum INFO:     ffmpeg-libs-2.6.4-1.fc22.x86_64 requires libhogweed.so.2()(64bit)
WARNING: potential problems with upgrade
  ffmpeg-libs-2.6.4-1.fc22.x86_64 (no replacement) requires nettle-2.7.1-5.fc22.x86_64 (replaced by nettle-3.1.1-3.fc23.x86_64)
  audit-libs-python3-2.4.3-1.fc23.x86_64 (new package) requires setroubleshoot-server-3.2.24-1.fc22.x86_64 (replaced by setroubleshoot-server-3.3.1-0.2.fc23.x86_64)
  python-solv-0.6.11-2.fc22.x86_64 (no replacement) requires libsolv-0.6.11-2.fc22.x86_64 (replaced by libsolv-0.6.11-3.git1f9abfb.fc23.x86_64)
fedup.yum INFO: checking for unneeded rpms in cache
fedup.yum INFO: beginning package download...
fedup.cli INFO: download GConf2-3.2.6-15.fc23.x86_64.rpm
warning: /var/cache/system-upgrade/fedora/packages/GConf2-3.2.6-15.fc23.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID 34ec9cba: NOKEY
fedup.yum WARNING: Public key for GConf2-3.2.6-15.fc23.x86_64.rpm is not installed
(1/2039): GConf2-3.2.6-15.fc23.x86_64.rpm                                        | 1.0 MB  00:00:00     
```

fedup wants to update 2039 packages and leave 3 not updated. (The total sum from system-upgrade does not match the total sum from fedup, I can dig deeper if needed and find the difference, but I think it's good enough for illustration). This looks as a much better and much safer outcome.

So, it seems that the "default behavior of the upgrade tool" already changed considerably. It seems that fedup simply ignored packages with broken deps and left them broken. That's probably inevitable during a system upgrade, the package set will never be 100% correct, especially with third-party apps. OTOH system-upgrade seems to try to keep the deps always satisfied, so it holds back a large part of the system packages just to make sure no deps are broken. That of course doesn't mean the upgraded system will work, it just means the RPM deps will not be broken. And it might be very likely that system-upgrade will reject to work at all if a suitable package with suitably broken deps is installed (which might still be a better outcome than to half-upgrade your system). A further problem is that system-upgrade doesn't tell you _why_ those packages are held back, so even a power user can't fix it.

What are we going to do about it? The first testers will start to use this once Beta is out on Sep 22, and our real end users will start to use this since Oct 27. We need to make sure we don't break half of the existing systems and this doesn't end up as a disaster. Currently I'm quite afraid of exactly that. I'm willing to spend considerable time testing this to avoid that.

My packaging experience is limited, I'd like to hear your opinions on the matter, but currently I think it's quite naive to use regular "dnf update" mechanism to perform a system upgrade. Even `apt` has `distro-upgrade` (and even `full-upgrade` since last year), to help avoid issues mentioned above. It allows to remove packages with broken deps if needed. It even says:

> apt-get has a "smart" conflict resolution system, and it will attempt to upgrade the most important packages at the expense of less important ones if necessary.

That sounds very appealing, though I'm not sure how well that works. It would certainly be great to see some logic like that in dnf/system-upgrade. But as a minimum, I think we need to allow to remove or ignore packages with broken deps, as fedup did. We could use `--allowerasing` for that, but in my quick testing it doesn't make any difference, so it might make sense to be used just with `distro-sync`. Can we perhaps use the same logic fedup used in system-upgrade as well?

> @kparal : The ffmpeg/vlc issue stems from rpmfusion not having rebuild their packages for f23:

Yes, that's exactly the real-world use case we need to make sure is handled reasonably well. Because none of our users run a clean, unaltered Fedora system. I'm quite glad we have something to test it easy with.

**Comment by jsilhan** [comment]
> `NameError: global name 'self' is not defined`

This is fixed in dnf-1.1.1-2

> Error: The operation would result in removing the booted kernel: kernel-core-4.1.6-200.fc22.x86_64.

Already reported bug in DNF. It can;t find it now but it was probably reported by `dmach` and has `kernel` in internal devel whiteboard.

> > apt-get has a "smart" conflict resolution system, and it will attempt to upgrade the most important packages at the expense of less important ones if necessary.
> 
> That sounds very appealing, though I'm not sure how well that works. It would certainly be great to see some logic like that in dnf/system-upgrade.

I'm working on global distro package preference approach and it could be handled by this as well.

> But as a minimum, I think we need to allow to remove or ignore packages with broken deps, as fedup did. We could use --allowerasing for that, but in my quick testing it doesn't make any difference, so it might make sense to be used just with distro-sync. Can we perhaps use the same logic fedup used in system-upgrade as well?

The use case described above is correct from DNF POV. You can either report conflict or succeed and remove the conficts. Othewise it's like installing via `--nodeps`. Skipping that package and let him there without satisfied deps is worse than removing it. DNF is in protected packages which requires systemd - you should be able to boot again.

**Comment by kparal** [comment]
I have reproduced the same issue with a pretty clean F22 VM. I just took almost default an up-to-date F22 installation, added rpmfusion, installed vlc, and tried to run system-upgrade. 1400 packages were updated, 230 packages were held back, including critical ones like `PackageKit`, `gnome-shell`, `gnome-software`, `gnome-terminal`, `grub2`, `mutter`, `nautilus`, `perl*`, `rpm`. All of that just because of vlc broken dep (!!). 

To my surprise, the upgraded system booted and seemed to work fine, even though F22 and F23 packages were mixed heavily. Not sure how lucky I was. Regular `dnf update` had no effect (as expected, it's the same logic as system-upgrade uses), but `dnf distro-sync --allowerasing` fixed the issue, it removed 25 packages (vlc and its deps) and upgraded/downgraded everything else. Great. 

Now we need to make sure this is done out of the box for all our users. Because rpmfusion or dropbox are notorious for being very late for new Fedora releases. And let's not even mention skype, which has no repo at all. The current approach could mean that 95% of our user base would end up with a half-upgraded system, and would hardly know how to fix it themselves.

> Skipping that package and let him there without satisfied deps is worse than removing it.

Agreed. The only thing I'm worried about is the removal logic in dnf. I could decide to keep vlc and remove the other half of the system. But I don't see a better approach than to work on improving that logic inside dnf/system-update if that occurs. E.g. prefer software from working repos to software from broken repos/no repos (in case it doesn't work that way already).

**Comment by kparal** [comment]
Digging deeper into the issues, the dependency problem is this:
`vlc` requires `ffmpeg` which requires `libhogweed.so.2()(64bit)` provided by `nettle`. `nettle` is so-bumped in F23 (`libhogweed.so.4()(64bit)`), thus breaking `ffmpeg` dependency. That's why dnf decided to keep old `nettle` and not update it. Of course, many standard packages depend on `nettle`, a recursively many other packages depend on those packages, so in the end about 230 packages are held back from updating.
`fedup` OTOH doesn't care about breaking `ffmpeg` dependency and updates all system packages, thus leaving `ffmpeg` broken but making the system fully upgraded.

What we need is to find such combination of `--best`, `--allowerasing`, `distro-sync` and new dnf/system-upgrade logic such that it will refuse to perform such a half-upgrade, but it will provide an option to remove offending packages and upgrade the rest of the system in full (prioritizing software from official fedora repos, if possible).

I have tried to reproduce a similar issue with other packages, and everything requiring ffmpeg (vlc, mplayer, etc) is affected at the moment. No other packages (skype, steam, dropbox, atom) were causing dep issues. But that's _at the moment_, and my experiments were of course limited. I think we're quite lucky we found out this ffmpeg issue right now, otherwise we wouldn't be aware of this issue. Now we can try to solve it before the new upgrade code is used in production.

@wgwoods, thoughts?

**Comment by jsilhan** [comment]
`--allowerasing` should be IMO used during dystro-sync. DNF should handle the fedora repo preference though. There are even more combinations to try:
`dnf repository-packages fedora remove-or-distro-sync --allowerasing`
`dnf repository-packages fedora remove-or-distro-sync \* --allowerasing`
`dnf distro-sync --allowerasing`
`dnf distro-sync \* --allowerasing`
`dnf distro-sync --disablerepo="*" --enablerepo=fedora \* --allowerasing`
`dnf distro-sync --disablerepo="*" --enablerepo=fedora --allowerasing`

**Comment by kparal** [comment]
I have managed to work around [rhbz 1260989](https://bugzilla.redhat.com/show_bug.cgi?id=1260989) by installing F23 kernel inside F22 and booting it. Then `distro-sync --allowerasing` started working just fine, and it seems to produce great results:
https://gist.github.com/kparal/45bc8b8d26e76c583ea7

As you can see it upgrades all packages to F23 (or downgrades, for those few that has a broken upgrade path), and removes those packages that would have broken deps (vlc, ffmpeg). Great. The removal operation seems a bit overzealous (it also removes packages without broken deps, like rpmfusion-free-release), I think that's [rhbz 1244175](https://bugzilla.redhat.com/show_bug.cgi?id=1244175).

I haven't tried the other proposed operations in previous comment, mainly because I think disabling all but official repos is not helpful (that would result in removing most of third-party content, even if the third-party repos contained an upgraded version). But if we find some issues with `distro-sync --allowerasing`, we can consider it.

Right now we are stuck on rhbz 1260989 and 1244175 (might be the same core issue). Once those are resolved, distro-sync could become the default, provided system-upgrade maintainers consider it a good solution. Or maybe there's a better solution up someone's sleeve. Please show us.

**Comment by kparal** [comment]
I have reported this as https://bugzilla.redhat.com/show_bug.cgi?id=1263677 and proposed as an F23 blocker. Opinions from system-upgrade and dnf folks would be very welcome.

**Comment by keszybz** [comment]
Implemented in 98ceee2211d677a8dfb5a33430dfa64dc8106d61.

**Comment by wgwoods** [comment]
Indeed.
